### PR TITLE
Elastic Beats: Set status.ObservedGeneration from metadata.Generation

### PR DESCRIPTION
**Pull request by naemono**
related to #3392 

This change allows status.ObservedGeneration to be kept in sync with metadata.Generation according to the rules described in the above issue. This change adds a set of unit tests which show more details into when observedGeneration will, and will not be updated according to any errors/situations that may occur during reconciliation. End to end tests were adjusted to always check for observedGeneration increments during any mutation changes.

### CODE REVIEW

**Comment by thbkrkr** [comment]
Jenkins test this please

**Comment by naemono** [comment]
I'm investigating an issue with this PR.  There seems to be another controller updating the status as soon as the beat controller updates the status, and there's a constant update loop between controllers.  Will update when resolved.

**Comment by thbkrkr** [comment]
> I'm investigating an issue with this PR. There seems to be another controller updating the status as soon as the beat controller updates the status, and there's a constant update loop between controllers. Will update when resolved.

Oh, this may be related to https://github.com/elastic/cloud-on-k8s/issues/5532.

File: pkg/controller/beat/common/reconcile.go [hunk 1]
```diff
@@ -71,13 +71,13 @@ func reconcilePodVehicle(podTemplate corev1.PodTemplateSpec, params DriverParams
...
 
-	err = updateStatus(params, ready, desired)
-	if err != nil && apierrors.IsConflict(err) {
-		params.Logger.V(1).Info(
-			"Conflict while updating status",
+	err = calculateStatus(params, ready, desired, params.Status)
+	if err != nil {
+		params.Logger.Error(
+			err,
+			"While calculating new status",
```
Comment (pebrc) [inline pkg/controller/beat/common/reconcile.go:78]: I would wrap the error to pass the context upwards and maybe not log here as controller runtime will log it anyway

**Review by pebrc — COMMENTED** [review]
Just a few nits, looks pretty good.

**Review by barkbay — COMMENTED** [review]
Almost LGTM, still trying to run the e2e tests.

File: pkg/controller/beat/common/reconcile.go [hunk 2]
```diff
@@ -133,21 +130,32 @@ func reconcileDaemonSet(rp ReconciliationParams) (int32, int32, error) {
 	return reconciled.Status.NumberReady, reconciled.Status.DesiredNumberScheduled, nil
 }
 
-func updateStatus(params DriverParams, ready, desired int32) error {
+// calculateStatus will calculate a new status from the state of the pods within the k8s cluster
+// and will return any error encountered.
+func calculateStatus(params DriverParams, ready, desired int32) error {
```
[the remark this answers is not in the document]
Comment (naemono) [reply pkg/controller/beat/common/reconcile.go]: I hopefully updated this to be consistent with your concerns.  If you wouldn't mind taking another look @barkbay I would appreciate it.  thanks.

**Pull request merged**